File: Array/35.c
#include<stdio.h>
#include<stdlib.h>

//search or insert position

int searchInsert(int* nums, int numsSize, int target) {
    int left = 0;
    int right = numsSize-1;
    int mid = 0;
    while(left <= right){
        mid = (left+right)/2;
        if(nums[mid] > target){
            right = mid -1;
        }
        else if(nums[mid]< target){
            left = mid +1;
        }
        else{
            return mid;
        }
    }
    if(left > right){
        return left;
    }
    return right;
}


int main(){
    int nums[] = {1,3,4,6};
    int numsSize = sizeof(nums) / sizeof(nums[0]);
    int target = 7;
    int result = searchInsert(nums, numsSize, target);
    printf("%d\n", result);

    return 0;
}

File: Array/35.py
class Solution:
    def searchInsert(self, nums, target):
        left = 0
        right = len(nums)-1

        while left <= right:
            mid = (left+right)//2
            if target > nums[mid]:
                left = mid+1
            elif target < nums[mid]:
                right = mid-1
            else:
                return mid
        return left


def run_test(nums, target, expected):
    sol = Solution()
    ans = sol.searchInsert(nums, target)
    print(f"nums={nums}, target={target}")
    print(f"  ans={ans}, expected={expected}  --> {'PASS' if ans == expected else 'FAIL'}")
    if ans != expected:
        raise AssertionError("Test failed")


if __name__ == "__main__":
    run_test([1,3,5,6], 5, 2)
    run_test([1,3,5,6], 2, 1)
    run_test([1,3,5,6], 7, 4)
    run_test([1,3,5,6], 0, 0)
    run_test([1], 0, 0)
    run_test([1], 2, 1)
    run_test([], 3, 0)
    print("All tests passed!")
